File: knowledgebase-gui/src/main/java/ru/usedesk/knowledgebase_gui/screens/pages/sections/SectionsAdapter.java
package ru.usedesk.knowledgebase_gui.screens.pages.sections;

import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;

import java.util.List;

import ru.usedesk.common_gui.external.IUsedeskViewCustomizer;
import ru.usedesk.knowledgebase_gui.R;
import ru.usedesk.knowledgebase_sdk.external.entity.Section;

public class SectionsAdapter extends RecyclerView.Adapter<SectionsAdapter.SectionViewHolder> {

    private final List<Section> sectionList;
    private final IOnSectionClickListener onSectionClickListener;
    private final IUsedeskViewCustomizer usedeskViewCustomizer;

    SectionsAdapter(@NonNull List<Section> sectionList,
                    @NonNull IOnSectionClickListener onSectionClickListener,
                    @NonNull IUsedeskViewCustomizer usedeskViewCustomizer) {
        this.sectionList = sectionList;
        this.onSectionClickListener = onSectionClickListener;
        this.usedeskViewCustomizer = usedeskViewCustomizer;
    }

    @NonNull
    @Override
    public SectionViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View view = usedeskViewCustomizer.createView(viewGroup, R.layout.usedesk_item_section);

        return new SectionViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull SectionViewHolder sectionViewHolder, int i) {
        sectionViewHolder.bind(sectionList.get(i));
    }

    @Override
    public int getItemCount() {
        return sectionList.size();
    }

    class SectionViewHolder extends RecyclerView.ViewHolder {

        private final View rootView;
        private final ImageView imageViewIcon;
        private final TextView textViewTitle;

        SectionViewHolder(@NonNull View itemView) {
            super(itemView);

            rootView = itemView;
            imageViewIcon = itemView.findViewById(R.id.iv_icon);
            textViewTitle = itemView.findViewById(R.id.tv_title);
        }

        void bind(@NonNull final Section section) {
            imageViewIcon.setImageBitmap(null);
            Glide.with(imageViewIcon)
                    .load(section.getImage())
                    .into(imageViewIcon);
            textViewTitle.setText(section.getTitle());

            rootView.setOnClickListener(v -> onSectionClickListener.onSectionClick(section.getId()));
        }
    }
}

File: knowledgebase-gui/src/main/java/ru/usedesk/knowledgebase_gui/screens/sections/SectionsAdapter.kt
package ru.usedesk.knowledgebase_gui.screens.sections

import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.lifecycle.LifecycleOwner
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import ru.usedesk.common_gui.UsedeskBinding
import ru.usedesk.common_gui.inflateItem
import ru.usedesk.common_gui.showImage
import ru.usedesk.knowledgebase_gui.R
import ru.usedesk.knowledgebase_sdk.entity.UsedeskSection

internal class SectionsAdapter(
    recyclerView: RecyclerView,
    viewModel: SectionsViewModel,
    lifecycleOwner: LifecycleOwner,
    private val onSectionClick: (Long, String) -> Unit
) : RecyclerView.Adapter<SectionsAdapter.SectionViewHolder>() {

    private var sections = listOf<UsedeskSection>()

    init {
        recyclerView.adapter = this
        viewModel.modelLiveData.initAndObserveWithOld(lifecycleOwner) { old, new ->
            if (old?.sections != new.sections) {
                val oldItems = sections
                val newItems = new.sections
                sections = newItems

                DiffUtil.calculateDiff(object : DiffUtil.Callback() {
                    override fun getOldListSize() = oldItems.size

                    override fun getNewListSize() = newItems.size

                    override fun areItemsTheSame(
                        oldItemPosition: Int,
                        newItemPosition: Int
                    ): Boolean {
                        val oldItem = oldItems[oldItemPosition]
                        val newItem = newItems[newItemPosition]
                        return oldItem.id == newItem.id
                    }

                    override fun areContentsTheSame(
                        oldItemPosition: Int,
                        newItemPosition: Int
                    ): Boolean {
                        val oldItem = oldItems[oldItemPosition]
                        val newItem = newItems[newItemPosition]
                        return oldItem.title == newItem.title &&
                                oldItem.thumbnail == newItem.thumbnail
                    }
                }).dispatchUpdatesTo(this)
            }
        }
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): SectionViewHolder {
        val binding = inflateItem(
            viewGroup,
            R.layout.usedesk_item_section,
            R.style.Usedesk_KnowledgeBase_Sections_Page_Section
        ) { rootView, defaultStyleId ->
            Binding(rootView, defaultStyleId)
        }
        return SectionViewHolder(binding)
    }

    override fun onBindViewHolder(sectionViewHolder: SectionViewHolder, i: Int) {
        sectionViewHolder.bind(sections[i])
    }

    override fun getItemCount(): Int = sections.size

    inner class SectionViewHolder(
        private val binding: Binding
    ) : RecyclerView.ViewHolder(binding.rootView) {

        private val noThumbnailId = binding.styleValues
            .getStyleValues(R.attr.usedesk_knowledgebase_sections_page_section_thumbnail_image)
            .getId(R.attr.usedesk_drawable_1)

        fun bind(section: UsedeskSection) {
            binding.ivIcon.setImageBitmap(null)
            binding.tvTitle.text = section.title
            binding.ivIcon.setImageResource(noThumbnailId)
            binding.tvInitials.text = section.title
            section.thumbnail?.also {
                showImage(binding.ivIcon,
                    noThumbnailId,
                    it,
                    onSuccess = {
                        binding.tvInitials.text = ""
                    }
                )
            }
            binding.lClickable.setOnClickListener {
                onSectionClick(section.id, section.title)
            }
        }
    }

    internal class Binding(
        rootView: View,
        defaultStyleId: Int
    ) : UsedeskBinding(rootView, defaultStyleId) {
        val lClickable: ViewGroup = rootView.findViewById(R.id.l_clickable)
        val ivIcon: ImageView = rootView.findViewById(R.id.iv_icon)
        val tvTitle: TextView = rootView.findViewById(R.id.tv_title)
        val tvInitials: TextView = rootView.findViewById(R.id.tv_initials)
    }
}